axum/src/response/sse.rs

```diff
@@ -451,11 +451,120 @@ impl KeepAliveStream {
     }
 }
 
-#[test]
-fn leading_space_is_not_stripped() {
-    let no_leading_space = Event::default().data("\tfoobar");
-    assert_eq!(no_leading_space.to_string(), "data: \tfoobar\n\n");
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{routing::get, test_helpers::*, Router};
+    use futures::{prelude::*, stream};
+    use std::{collections::HashMap, convert::Infallible};
+
+    #[test]
+    fn leading_space_is_not_stripped() {
+        let no_leading_space = Event::default().data("\tfoobar");
+        assert_eq!(no_leading_space.to_string(), "data: \tfoobar\n\n");
+
+        let leading_space = Event::default().data(" foobar");
+        assert_eq!(leading_space.to_string(), "data:  foobar\n\n");
+    }
+
+    #[tokio::test]
+    async fn basic() {
+        let app = Router::new().route(
+            "/",
+            get(|| async {
+                let stream = stream::iter(vec![
+                    Event::default().data("one").comment("this is a comment"),
+                    Event::default()
+                        .json_data(serde_json::json!({ "foo": "bar" }))
+                        .unwrap(),
+                    Event::default()
+                        .event("three")
+                        .retry(Duration::from_secs(30))
+                        .id("unique-id"),
+                ])
+                .map(Ok::<_, Infallible>);
+                Sse::new(stream)
+            }),
+        );
+
+        let client = TestClient::new(app);
+        let mut res = client.get("/").send().await;
+
+        assert_eq!(res.headers()["content-type"], "text/event-stream");
+        assert_eq!(res.headers()["cache-control"], "no-cache");
+
+        let event_fields = parse_event(&res.chunk_text().await.unwrap());
+        assert_eq!(event_fields.get("data").unwrap(), "one");
+        assert_eq!(event_fields.get("comment").unwrap(), "this is a comment");
+
+        let event_fields = parse_event(&res.chunk_text().await.unwrap());
+        assert_eq!(event_fields.get("data").unwrap(), "{\"foo\":\"bar\"}");
+        assert!(event_fields.get("comment").is_none());
+
+        let event_fields = parse_event(&res.chunk_text().await.unwrap());
+        assert_eq!(event_fields.get("event").unwrap(), "three");
+        assert_eq!(event_fields.get("retry").unwrap(), "30000");
+        assert_eq!(event_fields.get("id").unwrap(), "unique-id");
+        assert!(event_fields.get("comment").is_none());
+
+        assert!(res.chunk_text().await.is_none());
+    }
+
+    #[tokio::test]
```

Comment:
If you change this to `#[tokio::test(start_paused = true)]` then we can use [`time::advance`](https://docs.rs/tokio/latest/tokio/time/fn.advance.html) to have the test run more deterministically and without any delay.

Reply:
Good idea! I've changed the keep alive tests!